src/coreclr/jit/gentree.cpp

```diff
@@ -24415,58 +24415,69 @@ GenTree* Compiler::gtNewSimdSumNode(var_types type, GenTree* op1, CorInfoType si
     GenTree*       tmp       = nullptr;
 
 #if defined(TARGET_XARCH)
-    assert(!varTypeIsByte(simdBaseType) && !varTypeIsLong(simdBaseType));
-    assert(simdSize != 64);
 
-    // HorizontalAdd combines pairs so we need log2(vectorLength) passes to sum all elements together.
-    unsigned vectorLength = getSIMDVectorLength(simdSize, simdBaseType);
-    int      haddCount    = genLog2(vectorLength);
+    if (simdSize == 64)
+    {
+        assert(IsBaselineVector512IsaSupportedDebugOnly());
+        GenTree* op1Dup = fgMakeMultiUse(&op1);
+        op1             = gtNewSimdGetUpperNode(TYP_SIMD32, op1, simdBaseJitType, simdSize);
+        op1Dup          = gtNewSimdGetLowerNode(TYP_SIMD32, op1Dup, simdBaseJitType, simdSize);
+        simdSize        = simdSize / 2;
+        op1             = gtNewSimdBinOpNode(GT_ADD, TYP_SIMD32, op1, op1Dup, simdBaseJitType, simdSize);
+    }
 
     if (simdSize == 32)
     {
-        // Minus 1 because for the last pass we split the vector to low / high and add them together.
-        haddCount -= 1;
+        assert(compIsaSupportedDebugOnly(InstructionSet_AVX2));
+        GenTree* op1Dup = fgMakeMultiUse(&op1);
+        op1             = gtNewSimdGetUpperNode(TYP_SIMD16, op1, simdBaseJitType, simdSize);
+        op1Dup          = gtNewSimdGetLowerNode(TYP_SIMD16, op1Dup, simdBaseJitType, simdSize);
+        simdSize        = simdSize / 2;
+        op1             = gtNewSimdBinOpNode(GT_ADD, TYP_SIMD16, op1, op1Dup, simdBaseJitType, simdSize);
+    }
 
-        if (varTypeIsFloating(simdBaseType))
+    assert(simdSize == 16);
+    if (varTypeIsFloating(simdBaseType))
+    {
+        GenTree* op1Dup = nullptr;
+        if (simdBaseType == TYP_FLOAT)
         {
-            assert(compIsaSupportedDebugOnly(InstructionSet_AVX));
-            intrinsic = NI_AVX_HorizontalAdd;
+            assert(compIsaSupportedDebugOnly(InstructionSet_SSE));
+            intrinsic = NI_SSE_Shuffle;
+            op1Dup    = fgMakeMultiUse(&op1);
+            op1 = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, op1Dup, gtNewIconNode(177), intrinsic, simdBaseJitType, 16);
+            op1Dup = fgMakeMultiUse(&op1Dup);
+            op1    = gtNewSimdBinOpNode(GT_ADD, TYP_SIMD16, op1, op1Dup, simdBaseJitType, simdSize);
```

Comment:
This could do with a comment elaborating on how `177` is shuffling the parameters. Notably this is `0b1011_0001`, which is `[0, 1, 2, 3] -> [1, 0, 3, 2]` (thus giving `[0 + 1, 1 + 0, 2 + 3, 3 + 2]`).

A similar comment explaining the next shuffle is also worthwhile IMO. `3` is `0b0011`, which for `double` will do `[0, 1] -> [1, 1]` (giving `[0 + 1, 1 + 1]`). While for `float` will do `[0, 1, 2, 3] -> [3, 0, 0, 0]` (giving `[0 + 3, 1 + 0, 2 + 0, 3 + 0]`, or otherwise the complete addition of `[(0 + 1) + (3 + 2), (1 + 0) + (0 + 1), (2 + 3) + (0 + 1), (3 + 2) + (0 + 1)]`.

I think it might actually be better to customize the icon node passed for each of these two paths so we continue doing the equivalent of a "broadcast" of results for the V128 case, as it will allows some other future optimizations to kick in.

If we choose `1` (`0b0001`) for `double`, then we get `[0, 1] -> [1, 0]`, which gives us `[0 + 1, 1 + 0]`.

If we can choose `78` (`0b0100_1110`) for `float`, then we get `[0, 1, 2, 3] -> [2, 3, 0, 1]`, which gives us `[0 + 2, 1 + 3, 2 + 0, 3 + 1]` or otherwise, the complete addition of `[(0 + 1) + (2 + 3), (1 + 0) + (3 + 2), (2 + 3) + (0 + 1), (3 + 2) + (1 + 0)]`

----------

An example of where this would be beneficial is that it means we can update this optimization to kick in more: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/morph.cpp#L10572-L10593

Which is to say, we can avoid the pattern of `Create(Sum(....))` by preserving it as just `Sum` for `V128` and optimizing it to `broadcastv128` for `V256/V512`.

Reply:
Re the following part-
'Which is to say, we can avoid the pattern of Create(Sum(....)) by preserving it as just Sum for V128 and optimizing it to broadcastv128 for V256/V512.'

Won't this be an issue with the 'shift+add' logic used for integer types?
I am trying to avoid using shuffles

Reply:
It would be, but we can always limit the optimization to just floating-point where a shuffle is required and where its trivial to ensure the "better" thing happens.

Reply:
Updated this